gcloud-java-storage/src/main/java/com/google/gcloud/storage/testing/FakeStorageRpc.java

```diff
@@ -0,0 +1,269 @@
+      throws StorageException {
+    potentiallyThrow(options);
+    return null;
+  }
+
+  @Override
+  public boolean delete(Bucket bucket, Map<Option, ?> options) throws StorageException {
+    return false;
+  }
+
+  @Override
+  public boolean delete(StorageObject object, Map<Option, ?> options) throws StorageException {
+    String key = fullname(object);
+    contents.remove(key);
+    return null != stuff.remove(key);
+  }
+
+  @Override
+  public BatchResponse batch(BatchRequest request) throws StorageException {
+    return null;
+  }
+
+  @Override
+  public StorageObject compose(Iterable<StorageObject> sources, StorageObject target,
+                               Map<Option, ?> targetOptions) throws StorageException {
+    return null;
+  }
+
+  @Override
+  public byte[] load(StorageObject storageObject, Map<Option, ?> options) throws StorageException {
+    String key = fullname(storageObject);
+    if (!contents.containsKey(key)) {
+      throw new StorageException(404, "File not found: " + key);
+    }
+    return contents.get(key);
+  }
+
+  @Override
+  public Tuple<String, byte[]> read(
+      StorageObject from, Map<Option, ?> options, long zposition, int zbytes)
+      throws StorageException {
+    potentiallyThrow(options);
+    String key = fullname(from);
+    if (!contents.containsKey(key)) {
+      throw new StorageException(404, "File not found: " + key);
+    }
+    long position = zposition;
+    int bytes = zbytes;
+    if (position < 0) {
+      position = 0;
+    }
+    byte[] full = contents.get(key);
+    if ((int) position + bytes > full.length) {
+      bytes = full.length - (int) position;
+    }
+    if (bytes <= 0) {
+      // special case: you're trying to read past the end
+      return Tuple.of("etag-goes-here", new byte[0]);
+    }
+    byte[] ret = new byte[bytes];
+    System.arraycopy(full, (int) position, ret, 0, bytes);
+    return Tuple.of("etag-goes-here", ret);
+  }
+
+  @Override
+  public String open(StorageObject object, Map<Option, ?> options) throws StorageException {
+    String key = fullname(object);
+    boolean mustNotExist = false;
+    for (Option option : options.keySet()) {
+      if (option instanceof StorageRpc.Option) {
+        // this is a bit of a hack, since we don't implement generations.
+        if ((StorageRpc.Option) option == Option.IF_GENERATION_MATCH
+            && ((Long) options.get(option)).longValue() == 0L) {
+          mustNotExist = true;
+        }
+      }
+    }
+    if (mustNotExist && stuff.containsKey(key)) {
+      throw new StorageException(new FileAlreadyExistsException(key));
+    }
+    stuff.put(key, object);
+
+    return fullname(object);
+  }
+
+  @Override
+  public void write(String uploadId, byte[] toWrite, int toWriteOffset, long destOffset,
+                    int length, boolean last) throws StorageException {
+    // this may have a lot more allocations than ideal, but it'll work.
```

Comment:
In GCS, files are only considered "written" (data and metadata) and availale for get/listing/... if they 
were closed/finalized. You can do the same by having `open` store content and info in a staging area and move them to `contents` and `stuff` when `last == true`.